### PR TITLE
Error checking

### PR DESCRIPTION
**Pull request by eanyanwu**
This pull request is mainly for @superpowers11 but @fallcat, you can review it too if you would like 👍 

#154 

🎉 One more month before this gets fully used 🎉 

_As usual please don't be intimidated by the GitHub-produced diffs. They exaggerate the amount of changes I made._

This pull request focuses on removing the `return null` expressions from the LoginService. I replace them with throwing Exceptions.  
I think this is a good idea because:
- This explicitly lets faculty/non-reslife staff know upfront that they can't access the system instead of showing them an ugly error page :/
- Throwing an exception with a message tells the developer exactly what failed, where and why. This is in contrast to returning null which will eventually throw an unhelpful `ArgumentNullException` down the line if something went wrong.
- Flattens logic i.e. No more nested if/else in the `LoginController`

Testing:
- Currently writing UI tests for this. I will be testing login attempts with valid, invalid, missing and staff credentials (e.g. 360.facultytest; lol I forgot this existed 🐐 )
- I have manually tested this by:
    - Successfully login in as myself.
    - Getting an error message when I try to log in with no credentials
    - Getting an error message when I try to log in with incorrect credentials.
    - Getting an error message when I try to log in as 360.facultytest

### CODE REVIEW

File: Phoenix/Controllers/LoginController.cs [hunk 1]
```diff
@@ -40,67 +40,63 @@ public ActionResult Authenticate(LoginViewModel loginViewModel)
...
+            catch(UserNotFoundException ex)
+            {
+                _ADContext.Dispose();
+                loginViewModel.ErrorMessage = "Oh dear, it seems that username or password is invalid.";
+                return View("Index", loginViewModel);
+            }
+            catch(ArgumentNullException ex)
+            {
+                _ADContext.Dispose();
+                loginViewModel.ErrorMessage = "Please make sure you enter in values for the username.";
```
Comment (superpowers11) [inline Phoenix/Controllers/LoginController.cs]: Don't we NOT want to tell the user why a login didn't work? I remember discussing that. So it should always say a generic message about "username or password is incorrect" on the user side, even if on our side the reason why the login failed looks different.

File: Phoenix/Services/AdminDashboardService.cs [hunk 2]
```diff
@@ -91,43 +91,8 @@ where sessions.Contains(sess.SESS_CDE) && buildings.Contains(rci.BuildingCode)
...
-                //        result.Add("Chase/Fulton");
-                //    }
-                //}
-
-                // Needs thought: all those hard-coded checks above ^^ seem like they will make the system very brittle
-                // In order to change which buildings are associated, it will require a source code change. And it will not
-                // be possible for a new rci type to be added for multiple buildings in the future.
-                // What I am thinking: Maybe we should make it a one-for-one relationship between buildings and RCI types
-                // Yes, it will be some duplication, but it allows the user more power/customization when we are gone.
-                // #maintainability
```
Comment (superpowers11) [inline Phoenix/Services/AdminDashboardService.cs]: Teehee 🐐

File: Phoenix/Services/LoginService.cs [hunk 3]
```diff
@@ -82,7 +88,7 @@ public string GenerateToken(string username, string id)
             // Right now just hardcode to true for test purposes
             bool isAdmin = false;
 
-
+            // Witll throw an exception if the user has no role within the system. The exceptions is caught in the controller.
```
Comment (superpowers11) [inline Phoenix/Services/LoginService.cs]: *Will throw

File: Phoenix/Exceptions/UserNotFoundException.cs [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+            : base(string.Format("The user {0} was not found.", usr), innerException)
+        {
+            user = usr;
+        }
+
+        public string NotFoundUser()
+        {
+            return user;
+        }
+    }
```
Comment (superpowers11) [inline Phoenix/Exceptions/UserNotFoundException.cs]: These exception classes don't seem to _do_ much themselves... does the benefit lie simply in having more clearly delineated exceptions, that _could_ do something more specific if desired?

Comment (eanyanwu) [inline Phoenix/Exceptions/UserNotFoundException.cs]: I think the benefit is the naming? So when I throw an exception, its a bit clearer what the problem is. 
If it is more confusing than helpful, I'll get rid of them. I was going for making exception throwing easy to read :)

Comment (superpowers11) [inline Phoenix/Exceptions/UserNotFoundException.cs]: It did make it easier to read in the controller code. I was just a little confused when I got to the classes and found they were basically empty.

Comment (eanyanwu) [inline Phoenix/Exceptions/UserNotFoundException.cs]: Yup, just thin wrappers around the `Exception` class.
It's only for the name.

File: Phoenix/Controllers/LoginController.cs [hunk 5]
```diff
@@ -40,67 +40,63 @@ public ActionResult Authenticate(LoginViewModel loginViewModel)
...
-                // if user did not enter a password, set it to empty string, since this value must be non-null
-                password = "";
+                _ADContext = loginService.ConnectToADServer();
             }
-
-            if (!ModelState.IsValid)
+            catch (Exception e)
             {
-                loginViewModel.ErrorMessage = "Invalid model state.";
+                loginViewModel.ErrorMessage = "There was a problem connection to the server. We are sorry. Please try again later or contact the project maintainer.";
```
Comment (eanyanwu) [inline Phoenix/Controllers/LoginController.cs]: Should I just remove the "or contact the project maintainer" part?

File: Phoenix/Exceptions/UserNotFoundException.cs (hunk 4 above)
Comment (eanyanwu) [inline Phoenix/Exceptions/UserNotFoundException.cs]: 👍  or 👎 ?

**Pull request merged**